xds/src/main/java/io/grpc/xds/internal/security/certprovider/CertificateProvider.java

```diff
@@ -147,9 +166,12 @@ protected CertificateProvider(DistributorWatcher watcher, boolean notifyCertUpda
   @Override
   public abstract void close();
 
-  /** Starts the cert refresh and watcher update cycle. */
+  /** Starts the async cert refresh and watcher update cycle. */
   public abstract void start();
 
+  /** Initializes the instance by loading certs and updating watchers. */
```

Comment:
Why is this needed now but was not needed before? Why can't this be part of "start"?

Reply:
Great question - Eric mentioned that `start()` must be non blocking because of Netty event loop convention. So the calling client is in charge on how exactly the FileWatcherCertificateProvider will be behaving.
I'll double check with Eric if this approach is acceptable or we can change it to smth else

Reply:
SGTM. Leaving open for Eric to comment.

Reply:
Hi @ejona86 , so after looking at how `FileWatcherCertificateProvider` is used, I added an `init()` method to be called by it's client to synchronously load everything from the files. The only place where it's used is https://github.com/grpc/grpc-java/blob/master/xds/src/main/java/io/grpc/xds/internal/security/certprovider/CertificateProviderStore.java#L135 which is a kind of registry with cached instances. 
However after looking at how `FileWatcherCertificateProvider` is implemented:

- all the file paths are set via constructor, so they can't be changed after instantiation
- scheduling the load from the files requires a separate call by a client - `start()`
I wonder what is the value of instantiated but not `started` FWSP? From my perspective I'd rather just call both `init()` and `start()` as last part of the constructor, so the instance is fully initialized and monitors the files once it's stored in the registry.
WDYT?

Reply:
Doing I/O in the constructor is often discouraged. The start() was purposefully not part of the constructor. And start() goes out of its way to be asynchronous because you can't block in CertificateProviderStore as it is called on the EventLoop. Don't change any of that.

As I said before, you probably need to update https://github.com/grpc/grpc-java/blob/0b2c17d0da84e92fbb104bd46ccb2f61860db04b/xds/src/main/java/io/grpc/xds/internal/security/certprovider/CertProviderSslContextProvider.java#L155 to wait until all values are available. That is what powers the Netty code that initializes the connection. https://github.com/grpc/grpc-java/blob/0b2c17d0da84e92fbb104bd46ccb2f61860db04b/xds/src/main/java/io/grpc/xds/internal/security/SecurityProtocolNegotiators.java#L207

Netty is asynchronous, and the code is all asynchronous so we don't do blocking I/O on Netty threads. You can't avoid the asynchronous aspect.

Reply:
Ok, I was able to get rid of sync parts I introduced by changing how `checkAndReloadCertificates()` is organized - PTAL